agent/session.go

```diff
@@ -90,18 +90,36 @@ func (s *session) start(ctx context.Context) error {
 		description.Hostname = s.agent.config.Hostname
 	}
 
-	sessionCtx, cancel := context.WithTimeout(ctx, dispatcherRPCTimeout)
-	stream, err := client.Session(sessionCtx, &api.SessionRequest{
-		Description: description,
-	})
-	cancel()
-	if err != nil {
-		return err
-	}
+	errChan := make(chan error, 1)
+	var (
+		msg    *api.SessionMessage
+		stream api.Dispatcher_SessionClient
+	)
+
+	// Need to run Session in a goroutine since there's no way to set a
+	// timeout for an individual Recv call in a stream.
+	go func() {
+		stream, err = client.Session(ctx, &api.SessionRequest{
```

Comment:
It's sorta scary to do this thing just with `time.After` because there is no cleanup.
Maybe at least we can have WithCancel and cancel it in a case of timeout.